### ISSUE TITLE
[pod-spec] prometheus_scrape_unit_address may be set to None

### ISSUE THREAD

**Issue by phvalguima** [canonical/prometheus-k8s-operator#262]
### Bug Description

For pod-spec charms, the MetricsEndpointProvider may return None to prometheus_scrape_unit_address on:
```
    def _set_unit_ip(self, _):
...
        for relation in self._charm.model.relations[self._relation_name]:
            relation.data[self._charm.unit]["prometheus_scrape_unit_address"] = str(
                self._charm.model.get_binding(relation).network.bind_address
            )
```

Juju will only return an IP when the application pod defined on the pod-spec starts. One possible fix is to check if the bind_address is set to None and raise an exception:
```
    def _set_unit_ip(self, _):
...
        if not self._charm.model.get_binding(relation).network.bind_address:
            raise PrometheusProviderBindAddressNotAvailable()
        for relation in self._charm.model.relations[self._relation_name]:
            relation.data[self._charm.unit]["prometheus_scrape_unit_address"] = str(
                self._charm.model.get_binding(relation).network.bind_address
            )
```

### To Reproduce

1. juju deploy ch:argo-controller
2. juju deploy ch:minio
3. juju deploy ch:prometheus-k8s --channel=latest/edge
4. juju add-relation argo-controller prometheus-k8s
Argo-controller will not yet be ready to spin up the application container as minio relation is not set yet, prometheus will receive prometheus_scrape_unit_address=None

### Environment

microk8s v1.21
juju v2.9.28

### Relevant log output

```shell
$ juju show-unit prometheus-k8s/0
prometheus-k8s/0:
  opened-ports: []
  charm: local:focal/prometheus-k8s-0
  leader: true
  relation-info:
  - endpoint: prometheus-peers
    related-endpoint: prometheus-peers
    application-data: {}
    local-unit:
      in-scope: true
      data:
        egress-subnets: 10.152.183.74/32
        ingress-address: 10.152.183.74
        private-address: 10.152.183.74
  - endpoint: metrics-endpoint
    related-endpoint: metrics-endpoint
    application-data:
      alert_rules: '{"groups": [{"name": "argo_62a27dd7-6426-4fc4-85ba-4470c63f53e0_argo-controller_argo-controller_unit_unavailable_alerts",
        "rules": [{"alert": "ArgoUnitIsUnavailable", "expr": "up < 1", "for": "0m",
        "labels": {"severity": "critical", "juju_model": "argo", "juju_model_uuid":
        "62a27dd7-6426-4fc4-85ba-4470c63f53e0", "juju_application": "argo-controller",
        "juju_charm": "argo-controller"}, "annotations": {"summary": "Argo unit {{
        $labels.juju_model }}/{{ $labels.juju_unit }} unavailable", "description":
        "The argo unit {{ $labels.juju_model }} {{ $labels.juju_unit }} is unavailable
        LABELS = {{ $labels }}\n"}}]}]}'
      scrape_jobs: '[{"metrics_path": "/metrics", "static_configs": [{"targets": ["*:9090"]}]}]'
      scrape_metadata: '{"model": "argo", "model_uuid": "62a27dd7-6426-4fc4-85ba-4470c63f53e0",
        "application": "argo-controller", "unit": "argo-controller/0", "charm_name":
        "argo-controller"}'
    related-units:
      argo-controller/0:
        in-scope: true
        data:
          prometheus_scrape_unit_address: None
          prometheus_scrape_unit_name: argo-controller/0
  provider-id: prometheus-k8s-0
  address: 10.1.56.200
```
```


### Additional context

_No response_

**Comment by balbirthomas** [comment]
I can confirm this bug.

As suggested deferring the `relation-joined` event in `MetricsEndpointProvider` may be a valid way to fix this issue. However the semantics of Operator Framework's `event.defer()` effectively implies that events generated after the deferred event may still execute prior to the completion of the deferred event. Hence I would like to explore a solution (if possible) that avoids deferring the event. 

In the case of side car charm the simplest fix would be to just not set a unit IP address if the workload container has not yet been started. This is because when the workload container starts it will receive a `pebble-ready` event in response to which the `MetricsEndpointProvider` will set unit IP as seen in its `__intit__()` method
```
for container_name in charm.unit.containers:
            self.framework.observe(
                charm.on[container_name].pebble_ready,
                self._set_unit_ip,
            )
```
This will trigger a `relation-changed` event for the `MetricsEndpointConsumer` and that should result in the correct scrape target. 

However a fix may still be needed to `MetricsEndpointConsumer` since it should not have set a scrape target with `none` as the IP address. Perhaps `MetricsEndpointConsumer` should validate the IP address it has received from `MetricsEndpointProvider`. This validation may be done in the `_relation_hosts()` method of `MetricsEndpointProvider`. It may be sufficient to log a warning and just drop any invalid address here. 

Finally the last part of the solution that does need some exploring is what additional events can `MetricsEndpointProvider` bind the `_set_scrape_job_spec()` to so that pod spec charms do forward updated scrape target information to Prometheus when their pods have started successfully. The selected event must be such that it does not cause a regression for sidecar charms and yet provides a valid unit address for pod-spec charms.

**Comment by mmanciop** [comment]
I don't think `relation_joined` will work reliably for pods that churn.
Also, about the following:

> when the workload container starts it will receive a pebble-ready event
in response to which the MetricsEndpointProvider will set unit IP as seen
in its __intit__() method

AFAIK Pod Spec charms do NOT have pebble inside, so no pebble_ready is
fired.

**Comment by balbirthomas** [comment]
> AFAIK Pod Spec charms do NOT have pebble inside, so no pebble_ready is fired.

Indeed, hence I mentioned "Finally the last part of the solution that does need some exploring is what additional events ...".

**Comment by mmanciop** [comment]
> > AFAIK Pod Spec charms do NOT have pebble inside, so no pebble_ready is fired.
> 
> Indeed, hence I mentioned "Finally the last part of the solution that does need some exploring is what additional events ...".

You see, I do not understand why would you change something that does work for workload containers that do have pebble ;-)

**Comment by balbirthomas** [comment]
> You see, I do not understand why would you change something that does work for workload containers that do have pebble ;-)

I think I am in total agreement with you here. I would not like to change the workflow for containers that have pebble. Hence I intend to explore if it is possible to bind *an additional event* that will not break the workflow for sidecar charms but additionally make it work for pod spec charms.

**Comment by simskij** [comment]
The only way we can reasonably solve this for PodSpec charms is to hook onto `update-status`, doing nothing until the IP is no longer `None`. Deferring is not a reasonable option, and there is no consistent event we can rely on that is even remotely equivalent to `pebble-ready`.

**Comment by phvalguima** [comment]
Hi, I've discussed this bug with @simskij. The main concern is to avoid deferring an event within the prometheus lib.

The two alternatives we discussed are: (1) as @simskij pointed out, the library could listen to more events, such as ```update-status``` or ```config-changed``` and use the ```_set_unit_ip``` to only set the IP when these events happen; or (2) create an event specific to no-ip-found, in this case, the main charm will have to listen to that event.

The main point with (2) is that it will demand an extra interface on the library, e.g. ```refresh``` method, that allows to inform the IP once it has been set.

**Comment by balbirthomas** [comment]
@phvalguima There is now [PR 268](https://github.com/canonical/prometheus-k8s-operator/pull/268) that hopefully address this issue. Any feedback you may provide regarding if this feedback effectively resolves your issue will be very helpful. I am waiting on integration tests to run and trying to follow the steps you provided in "Steps to reproduce" locally now.

**Comment by balbirthomas** [comment]
@phvalguima I am unable to add a relation between `argo-controller` (deployed from charmhub edge channel) and `prometheus-k8s` build using PR #268.
```
$ juju deploy argo-controller --channel edge
Located charm "argo-controller" in charm-hub, revision 59
Deploying "argo-controller" from charm-hub charm "argo-controller", revision 59 in channel edge on focal
ERROR cannot add application "argo-controller": state changing too quickly; try again soon
```
Is there any particular revision you would suggest ? Stable and beta do not seem to support the Prometheus charm's `metrics-endpoint` relation.
```
$ juju relate argo-controller prometheus-k8s
ERROR no relations found
``

**Comment by balbirthomas** [comment]
Hi @phvalguima I can confirm PR #268 does indeed fix this issue. Here is the `juju status`

```
$ juju status --relations
Model  Controller  Cloud/Region        Version  SLA          Timestamp
lma    microk8s    microk8s/localhost  2.9.29   unsupported  13:28:59+01:00

App              Version                Status   Scale  Charm            Channel  Rev  Address         Exposed  Message
argo-controller                         blocked      1  argo-controller             0                  no       Waiting for object-storage relation data
minio            res:oci-image@1755999  active       1  minio            stable    57  10.152.183.119  no       
prometheus-k8s                          active       1  prometheus-k8s              0  10.152.183.44   no       

Unit                Workload  Agent  Address      Ports     Message
argo-controller/0*  blocked   idle                          Waiting for object-storage relation data
minio/0*            active    idle   10.1.252.84  9000/TCP  
prometheus-k8s/0*   active    idle   10.1.252.85            

Relation provider                 Requirer                         Interface          Type     Message
argo-controller:metrics-endpoint  prometheus-k8s:metrics-endpoint  prometheus_scrape  regular  
prometheus-k8s:prometheus-peers   prometheus-k8s:prometheus-peers  prometheus_peers   peer     
```

And here is the `juju show-unit` output

```
$ juju show-unit argo-controller/0
argo-controller/0:
  opened-ports: []
  charm: local:kubernetes/argo-controller-0
  leader: true
  relation-info:
  - relation-id: 1
    endpoint: metrics-endpoint
    related-endpoint: metrics-endpoint
    application-data: {}
    related-units:
      prometheus-k8s/0:
        in-scope: true
        data:
          egress-subnets: 10.152.183.44/32
          ingress-address: 10.152.183.44
          private-address: 10.152.183.44
```
Also there is no scrape target for `argo-controller` in the Prometheus config file
```
global:
  scrape_interval: 1m
  scrape_timeout: 10s
  evaluation_interval: 1m
rule_files:
- /etc/prometheus/rules/juju_*.rules
scrape_configs:
- job_name: prometheus
  honor_timestamps: true
  scrape_interval: 5s
  scrape_timeout: 5s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  static_configs:
  - targets:
    - localhost:9090
- job_name: juju_lma_938fb387-bf33-48b8-8ead-3daa04fc7c94_argo-controller_argo-controller_prometheus_scrape
  honor_labels: true
  honor_timestamps: true
  scrape_interval: 1m
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [juju_model, juju_model_uuid, juju_application]
    separator: _
    regex: (.*)
    target_label: instance
    replacement: $1
    action: replace
```

After adding a relation between `argo-controller` and `minio` , the Prometheus charm then does get a valid scrape target for `argo-controller`.  Screenshot is attached. Note however in the attached screenshot Prometheus is still not able to scrape from the argo controller because it is getting a "permission denied" error.
![Screenshot from 2022-05-04 13-33-18](https://user-images.githubusercontent.com/1672069/166682418-50a4e716-4fb5-4ba0-b193-d80e4fcaaf23.png)
